fallback/s2a_fallback.go

```diff
@@ -0,0 +1,159 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package fallback provides default implementations of fallback options when S2A fails.
+package fallback
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/grpclog"
+	"net"
+)
+
+const (
+	alpnProtoStrH2   = "h2"
+	alpnProtoStrHttp = "http/1.1"
+	defaultHttpsPort = "443"
+)
+
+// FallbackClientHandshake establishes a TLS connection and returns it, plus its auth info.
+// Input:
+//
+//	targetServer: the server attempted with S2A.
+//	conn: the raw tcp connection passed into S2A's ClientHandshake func.
+//	            If fallback is successful, the `conn` should be closed.
+//	err: the error encountered when performing client handshake with S2A.
+type FallbackClientHandshake func(ctx context.Context, targetServer string, conn net.Conn, err error) (net.Conn, credentials.AuthInfo, error)
+
+// DefaultFallbackClientHandshakeFunc returns an implementation of FallbackOptions.FallbackClientHandshakeFunc,
+// which is of type FallbackClientHandshake.
+// It establishes a TLS connection with the provided fallbackAddr, returns the new connection and its auth info.
+// Example use:
+//
+//	transportCreds, _ = s2a.NewClientCreds(&s2a.ClientOptions{
+//		S2AAddress: s2aAddress,
+//		EnableV2:   true,
+//		FallbackOpts: &s2a.FallbackOptions{ // optional
+//			FallbackClientHandshakeFunc: fallback.DefaultFallbackClientHandshakeFunc(fallbackAddr),
+//		},
+//	})
+//
+// The fallback server's certificate should be verifiable using OS root store.
+// The fallbackAddr is expected to be a network address, e.g. example.com:port. If port is not specified,
+// it uses default port 443.
+// In the returned function's TLS config, ClientSessionCache is explicitly set to nil to disable TLS resumption,
+// and min TLS version is set to 1.3.
+func DefaultFallbackClientHandshakeFunc(fallbackAddr string) (func(context.Context, string, net.Conn, error) (net.Conn, credentials.AuthInfo, error), error) {
+	fallbackServerAddr, err := processFallbackAddr(fallbackAddr)
+	if err != nil {
+		if grpclog.V(1) {
+			grpclog.Infof("error processing fallback address [%s]: %v", fallbackAddr, err)
+		}
+		return nil, err
+	}
+	return func(ctx context.Context, targetServer string, conn net.Conn, s2aErr error) (net.Conn, credentials.AuthInfo, error) {
+		fallbackTLSConfig := tls.Config{
+			MinVersion:         tls.VersionTLS13,
+			ClientSessionCache: nil,
+			NextProtos:         []string{alpnProtoStrH2},
+		}
+		fallbackDialer := &tls.Dialer{Config: &fallbackTLSConfig}
+		fbConn, fbErr := fallbackDialer.DialContext(ctx, "tcp", fallbackServerAddr)
```

Comment:
nit: the usual google3 Go convention is to use the variable name `err` for all errors.

Reply:
in this case, there's an `s2aErr` from s2a failure, which is passed into the func, then `fbErr` is from dialing the fallback connection, I'd like the naming to clearly differentiate the two.

Reply:
SGTM, makes sense.